command/agent/consul/connect.go

```diff
@@ -226,38 +233,59 @@ func connectMeshGateway(in structs.ConsulMeshGateway) api.MeshGatewayConfig {
 	return gw
 }
 
-func connectProxyConfig(cfg map[string]interface{}, port int, allocID string) map[string]interface{} {
+func connectProxyConfig(cfg map[string]interface{}, port int, info allocInfo) map[string]interface{} {
 	if cfg == nil {
 		cfg = make(map[string]interface{})
 	}
 	cfg["bind_address"] = "0.0.0.0"
 	cfg["bind_port"] = port
-	injectAllocID(cfg, allocID)
+
+	tags := map[string]string{
+		"nomad.group=":     info.Group,
+		"nomad.job=":       info.JobID,
+		"nomad.namespace=": info.Namespace,
+		"nomad.alloc_id=":  info.AllocID,
+	}
+	injectNomadInfo(cfg, tags)
 	return cfg
 }
 
-// injectAllocID merges allocID into cfg=>envoy_stats_tags
+// injectNomadInfo merges nomad information into cfg=>envoy_stats_tags
 //
 // cfg must not be nil
-func injectAllocID(cfg map[string]interface{}, allocID string) {
+func injectNomadInfo(cfg map[string]interface{}, defaultTags map[string]string) {
 	const key = "envoy_stats_tags"
-	const prefix = "nomad.alloc_id="
-	pair := prefix + allocID
 	tags, exists := cfg[key]
 	if !exists {
```

Comment:
We're only checking for existence here and not that it's the expected shape, so we end up falling thru entirely if it's not. Maybe we should do the type switch here and also overwrite if len==0 ? That'll be the most common case by far and will save us from entering the more complex loop below.

Reply:
Tried to do this, but I don't know if I understood you correctly. One thing that's changed respect to the previous behavior is that right now if it's not a slice the value would be overridden by the nomad default tags. Should we leave the value provided by the user in that case?